### DISCUSSION TITLE
Vue2 Accuracy

### DISCUSSION THREAD

**Discussion by oliv3r** [emporia-vue-local/esphome#58] [Ideas]
Hey all,

I've been eying the Vue2 a lot lately, however I ran into this comment 'on the interwebs' (can't find it of course anymore) that was saying that the CT's shipped with the Vue2 are poor and the ADC of the atsam weak. The CT's I've read that elsewhere, that the accuracy is not that great, but that is easily resolved by using better/different CT's (I want smaller ones for example, without clamps). But the ADC doesn't seem to horrible. The atmel sam09du is a 12bit ADC, which can even oversample to 16bit with intermediate 22bit resolution. So unless PCB design is not ideal, and the atmel MCU software could be improved, we should be able to measure quite nicely.

Going down the rabbit hole and re-writing the atsam firmware (lets hope it's not write-protected) would make it even better.

Thoughts?

**Comment by flaviut** [comment]
> The CT's I've read that elsewhere, that the accuracy is not that great ... But the ADC doesn't seem to horrible

I've had no problem with accuracy. A calibration constant for voltage is needed for some reason depending on the order of the phases, but my Vue's readings are within a few percent of the power company's readings.

> lets hope it's not write-protected

Didn't even realize that was possible! I knew about read-protection, but that'd really suck. On the other hand, I came across this project just a bit ago: https://github.com/KULeuven-COSIC/Starlink-FI. It's interesting how little it takes to bypass these "security" measures!

> Thoughts?

It'd be really interesting to see a rewrite of the atsam firmware! I'd expect the current thing is good enough for 90% of users, [but there's always that last 10%](https://github.com/emporia-vue-local/esphome/discussions/63).

**Reply by oliv3r** [reply]
> within a few percent of the power company's readings.

That doesn't give confidence though; I've read 2% accuracy, but I'd be hoping to get much better. So then the question is, is that because of (lack of) proper calibration, or that's just the best the hardware can do ...

> Didn't even realize that was possible! 

Well depends on the part I suppose, but if they are doing high-volumes, having a masked rom is an option. I wonder if they did a i2c bootloader though; so they can do firmware updates ota of the atsam ...
